### DISCUSSION TITLE
C++ build fail - Header files in include path not found - c_cpp_properties.json settings invoking compiler?

### DISCUSSION THREAD

**Discussion by AAnthon1985** [microsoft/vscode-cpptools#7047] [Q&A] [answered]
Hi,
before I start my thread here, I would like to mention that I asked the same question on stackoverflow:
https://stackoverflow.com/questions/66354203/vscode-mingw-g-header-files-include-fail-from-subdirectory

The reason that I am asking this question here again is that I read the description and the tutorial on the vscode website differently to the answer I received on stackoverflow, so I would like to have it confirmed that I am wrong (or perhaps right). I've always believed that my c_cpp_properties.json file can change the arguments when invoking my compiler. So if I want to have a particular include path, I put it into my c_cpp_properties.json file which then gives the information further to my tasks.json where to look for the header file. Am I correct or wrong in my understanding?

To my actual problem:

Following the guideline on how to setup mingw g++ in vscode, I cannot seem to succeed in including headerfiles from a subdirectory. I have my working folder, in which my main.cpp lies. In that working folder, I have a subfolder called "inc" in which my header file add.h lies, and I have a subfolder "src" in which my add.cpp file lies. When I pass the argument to the compiler to include header files using the -I argument, it works fine. But this kind of contradicts the tutorial found on the vscode website that states that I should add include paths in the c_cpp_properties.json file:
https://code.visualstudio.com/docs/cpp/config-mingw

The build output message:

> C:\MinGW\bin\g++.exe -g C:\Users\Alexander\Documents\GitHub\CPP_Projects\00_helloworld_min*.cpp C:\Users\Alexander\Documents\GitHub\CPP_Projects\00_helloworld_min\src*.cpp -o C:\Users\Alexander\Documents\GitHub\CPP_Projects\00_helloworld_min\main.exe C:\Users\Alexander\Documents\GitHub\CPP_Projects\00_helloworld_min\main.cpp:2:17: fatal error: add.h: No such file or directory #include "add.h"

A simple minimalistic code:
```#include <iostream>
#include "add.h"
 
int main()
{
    std::cout << "The sum of 15 and 5 is " << add(15, 5) << std::endl;
    return 0;
}
```
A screenshot of my project to illustrate my inc and src folder:

![vscode](https://user-images.githubusercontent.com/68788787/109173543-3d22a080-7784-11eb-809a-560b3526e952.PNG)

My tasks.json:
```
{
    "version": "2.0.0",
    "tasks": [
        {
            "type": "cppbuild",
            "label": "C/C++: g++.exe build active file",
            "command": "C:\\MinGW\\bin\\g++.exe",
            "args": [
                "-g",
                "${workspaceFolder}\\*.cpp",
                "${workspaceFolder}\\src\\*.cpp",
                // "-I",
                // "${workspaceFolder}\\inc\\",
                "-o",
                "${fileDirname}\\${fileBasenameNoExtension}.exe"
            ],
            "options": {
                "cwd": "${workspaceFolder}"
            },
            "problemMatcher": [
                "$gcc"
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "detail": "compiler: C:\\MinGW\\bin\\g++.exe"
        }
    ]
}
```

My c_cpp_properties.json:
```
{
    "configurations": [
        {
            "name": "GCC",
            "includePath": [
                "${workspaceFolder}\\**",
                "${workspaceFolder}\\inc\\**"
            ],
            "defines": [
                "_DEBUG",
                "UNICODE",
                "_UNICODE"
            ],
            "compilerPath": "C:\\MinGW\\bin\\gcc.exe",
            "cStandard": "gnu11",
            "cppStandard": "gnu++14",
            "intelliSenseMode": "windows-gcc-x86"
        }
    ],
    "version": 4
}
```

Thank you,

**Comment by sean-mcmanus** [comment, marked as answer]
tasks.json and c_cpp_properties.json are separate with no interaction between them, so both need to be configured separately (usually in the "equivalent" way).

**Reply by AAnthon1985** [reply]
Hi @sean-mcmanus 
Thank you for the reply and explanation. That confirms the answer that I received on stackoverflow, and I will have peace with it :)

Just so that I understand it:
In my tasks.json file, I set the arguments of my compiler, for instance where to look for header files, which C++ standard I want to compile with, which optimization level I want to have etc. The tasks.json is important that defines the outcome of my compilation.
What exactly is my c_cpp_properties.json file then doing? Is it "just" setting the behavior of my intellisense to give me the best possible user programming experience to be as close as possible to my actual compiler settings, but has nothing to do with the actual compilation?

Thank you very much,

**Reply by sean-mcmanus** [reply]
Yes, the c_cpp_properties.json settings, compilerPath, compilerArgs, etc. are just used for configuring IntelliSense. When a configurationProvider is set and an extension like CMake Tools or Makefile Tools is used then the build and IntelliSense configuration information is shared (provided by the configuration provider), but we haven't implemented any way for cppbuild tasks to provide configuration information to IntelliSense (or the opposite direction).

**Reply by AAnthon1985** [reply]
Thank you very much for the detailed answer.  This topic #7047 is answered and can be closed from my side. Much appreciated. Cheers,
